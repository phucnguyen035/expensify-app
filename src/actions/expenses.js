import uuidv4 from 'uuid/v4';
import db from '../database/firebase';

// ADD_EXPENSE
const addExpense = newExpense => ({
  type: 'ADD_EXPENSE',
  expense: newExpense
});

const startAddExpense = (expenseData = {}) => (dispatch) => {
  const { amount, createdAt, desc, note } = expenseData;
  const newExpense = { amount, createdAt, desc, note };

  db.ref('expenses')
    .push(newExpense)
    .then((ref) => {
      dispatch(
        addExpense({
          id: ref.key,
          ...newExpense
        })
      );
    });
};

// REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export { addExpense, startAddExpense, removeExpense, editExpense };
