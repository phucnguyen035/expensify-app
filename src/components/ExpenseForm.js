import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      desc: props.desc ? props.desc : '',
      note: props.note ? props.note : '',
      amount: props.amount ? (props.amount / 100).toString() : '',
      createdAt: props.createdAt ? moment(props.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  onDescChange = (e) => {
    const desc = e.target.value;

    this.setState(() => ({ desc }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // Match amount with number only with 2 decimal places
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused }));
  };

  onFormSubmit = (e) => {
    e.preventDefault();
    if (!this.state.desc || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        desc: this.state.desc,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onFormSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input
          type="text"
          className="text-input"
          value={this.state.desc}
          placeholder="Description"
          onChange={this.onDescChange}
        />

        <input
          type="text"
          className="text-input"
          value={this.state.amount}
          placeholder="Amount"
          onChange={this.onAmountChange}
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.focused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          displayFormat="D/M/YYYY"
        />

        <textarea
          className="text-area"
          value={this.state.note}
          onChange={this.onNoteChange}
          placeholder="Add a note for this expense (optional)"
        />

        <div>
          <button className="button">Save</button>
        </div>
      </form>
    );
  }
}

ExpenseForm.propTypes = {
  desc: PropTypes.string,
  amount: PropTypes.number,
  note: PropTypes.string,
  onSubmit: PropTypes.func,
  createdAt: PropTypes.number
};

ExpenseForm.defaultProps = undefined;
