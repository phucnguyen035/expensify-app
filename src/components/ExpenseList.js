import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import getVisibleExpenses from '../selectors/expenses';
import { ExpenseListItem } from './ExpenseListItem';

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

export const ExpenseList = props => (
  <div className="container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>

    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item list-item--no-data">
          <span>No expenses</span>
        </div>
      ) : (
        props.expenses.map(expense => <ExpenseListItem {...expense} key={expense.id} />)
      )}
    </div>
  </div>
);

ExpenseList.propTypes = {
  expenses: PropTypes.arrayOf(PropTypes.any)
};

ExpenseList.defaultProps = undefined;

export default connect(mapStateToProps)(ExpenseList);
