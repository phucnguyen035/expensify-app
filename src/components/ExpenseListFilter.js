import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilters,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';

const mapStateToProps = state => ({
  filters: state.filters
});
const mapDispatchtoProps = dispatch => ({
  setTextFilters: text => dispatch(setTextFilters(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export class ExpenseListFilter extends Component {
  state = {
    focusedInput: null
  };

  // Sort function
  onSortChange = (sortValue) => {
    if (sortValue === 'date') {
      this.props.sortByDate();
    } else if (sortValue === 'amount') {
      this.props.sortByAmount();
    }
  };

  onTextChange = (text) => {
    this.props.setTextFilters(text);
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  focusedInput = (focusedInput) => {
    this.setState(() => ({ focusedInput }));
  };

  render() {
    return (
      <Fragment>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => this.onTextChange(e.target.value)}
        />
        <select value={this.props.filters.sortBy} onChange={e => this.onSortChange(e.target.value)}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate}
          // startDateId="startDate_filters"
          endDate={this.props.filters.endDate}
          // endDateId="endDate_filters"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.focusedInput}
          isOutsideRange={() => false}
          numberOfMonths={1}
          showClearDates
          displayFormat="DD/MM/YYYY"
        />
      </Fragment>
    );
  }
}

ExpenseListFilter.propTypes = {
  filters: PropTypes.objectOf(PropTypes.any),
  sortByDate: PropTypes.func,
  sortByAmount: PropTypes.func,
  setTextFilters: PropTypes.func,
  setStartDate: PropTypes.func,
  setEndDate: PropTypes.func
};

ExpenseListFilter.defaultProps = undefined;

export default connect(
  mapStateToProps,
  mapDispatchtoProps
)(ExpenseListFilter);
