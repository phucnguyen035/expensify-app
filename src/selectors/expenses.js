import moment from 'moment';
// Visible expenses
export default (expenses, { text, startDate, endDate, sortBy }) =>
  expenses
    .filter((expense) => {
      const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt)) : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt)) : true;
      const textMatch = expense.desc.toLowerCase().includes(text.toLowerCase());

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
      // Default return latest item
      return a.createdAt < b.createdAt ? 1 : -1;
    });
