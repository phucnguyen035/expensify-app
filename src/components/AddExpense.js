import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export class AddExpense extends Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="container">
            <h2 className="page-header__title">Add expense</h2>
          </div>
        </div>

        <div className="container">
          <button className="button button--back" onClick={this.props.history.goBack}>
            Back
          </button>
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

AddExpense.propTypes = {
  startAddExpense: PropTypes.func,
  history: PropTypes.shape({
    goBack: PropTypes.func,
    push: PropTypes.func
  }).isRequired
};

AddExpense.defaultProps = undefined;

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpense);
