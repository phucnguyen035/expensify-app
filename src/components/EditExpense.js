import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});
const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export class EditExpense extends Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/dashboard');
  };
  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>

        <div className="container">
          <button className="button button--back" onClick={this.props.history.goBack}>
            Back
          </button>

          <ExpenseForm {...this.props.expense} onSubmit={this.onSubmit} />

          <button className="button button--secondary" onClick={this.onClick}>
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

EditExpense.propTypes = {
  startEditExpense: PropTypes.func.isRequired,
  startRemoveExpense: PropTypes.func.isRequired,

  history: PropTypes.shape({
    goBack: PropTypes.func,
    push: PropTypes.func
  }).isRequired,

  expense: PropTypes.objectOf(PropTypes.any).isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
