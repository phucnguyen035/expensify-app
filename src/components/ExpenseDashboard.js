import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
