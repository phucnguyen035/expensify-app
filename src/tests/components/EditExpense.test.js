import React from 'react';
import { EditExpense } from '../../components/EditExpense';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

describe('Edit expense page', () => {
  let wrapper, history, editExpense, removeExpense;

  beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
      <EditExpense
        expense={expenses[0]}
        history={history}
        editExpense={editExpense}
        removeExpense={removeExpense}
      />
    );
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle editExpense onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenCalledWith(expenses[0].id, expenses[0]);
  });

  test('should handle removeExpense onClick', () => {
    wrapper.find('button').simulate('click');

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[0].id });
  });
});
