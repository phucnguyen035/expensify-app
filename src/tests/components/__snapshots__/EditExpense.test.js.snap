// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit expense page should render correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="container"
  >
    <button
      className="button button--back"
    >
      Back
    </button>
    <ExpenseForm
      amount={500}
      createdAt={0}
      desc="test 1"
      id="1"
      note="test 1 note"
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
